libraries/lib-music-information-retrieval/tests/MirTestUtils.cpp

```diff
@@ -0,0 +1,83 @@
+#include "MirTestUtils.h"
+
+#define USE_FILESYSTEM (__has_include(<filesystem>) && _WIN32)
+
+#if USE_FILESYSTEM
+#   include <filesystem>
+#endif
+
+#include <array>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+
+namespace MIR
+{
+std::vector<std::string> GetWavFilesUnderDir(const char* dir)
+{
+   // Only compile this if std::filesystem is available.
+   std::vector<std::string> files;
+#if USE_FILESYSTEM
+   namespace fs = std::filesystem;
+   for (const auto& entry : fs::directory_iterator(dir))
+      // Ignore directories starting with underscore.
+      if (entry.is_directory() && entry.path().filename().string()[0] != '_')
+      {
+         // Look for a file `audacity-maxNumFiles.txt` in this directory. If it
+         // exists and contains a number, only process that many files.
+         const auto maxNumFilesFilename =
+            entry.path().string() + "/audacity-maxNumFiles.txt";
+         std::ifstream maxNumFilesFile { maxNumFilesFilename };
+         size_t maxNumFiles = std::numeric_limits<size_t>::max();
+         if (maxNumFilesFile.good())
+         {
+            std::string line;
+            std::getline(maxNumFilesFile, line);
+            maxNumFiles = std::stoi(line);
+         }
+         // Collect up to `maxNumFiles` wav files in this directory.
+         size_t numFiles = 0;
+         for (const auto& subEntry : fs::recursive_directory_iterator(entry))
+            if (
+               subEntry.is_regular_file() &&
+               subEntry.path().extension() == ".wav")
+            {
+               files.push_back(subEntry.path().string());
+               if (++numFiles >= maxNumFiles)
+                  break;
+            }
+      }
+#endif
+   return files;
+}
+
+void ProgressBar(int width, int percent)
+{
+   int progress = (width * percent) / 100;
+   std::cout << "[";
+   for (int i = 0; i < width; ++i)
+      if (i < progress)
+         std::cout << "=";
+      else
+         std::cout << " ";
+   std::cout << "] " << std::setw(3) << percent << "%\r";
+   std::cout.flush();
+}
+
+OctaveError GetOctaveError(double expected, double actual)
```

Comment:
I or another maintainer would need help understanding this uncommented function.

I think I understand, from examining usage, that expected and actual really are frequencies, but not of audible tones -- rather the low frequencies of beats per minute.

"expected" might perhaps be called "nominal" as the bpm parsed from a filename.  "actual" might perhaps be called "measured" as by your analyses.

The function finds a discrepancy between them on a logarithmic scale.  (Really any logarithmic base could do as well as 2 for the purpose of ranking.)

Except that it allows displacing one of them on the scale by the listed five factors, whichever minimizes error.  I don't quite get that part.

How should I interpret it when the chosen factor is not 1?

I look through your csv files for such examples.  They are a minority of the positive population.  Only one with 1/3, three with 1/2, one with 3.  2 is more frequent.

Reply:
I think I should understand this allowance of factors as the "common denominator" thing?  You have explained to me how the tatums that the analysis detects might really be more or fewer than the quarter note beats, by various whole or fractional multipliers.

Reply:
... and yet `possibleTatumsPerBeat` in `GetMeterTatumUsingQuantizationFit.cpp` lists nine possibilities.

So 4, 6, 1/4, and 1/6 are the even more outlying possibilities, of which, you don't yet have examples in your dataset?

Reply:
The best explanation you can get is probably in the paper I got this from, [Music Tempo Estimation: Are We Done Yet?](https://storage.googleapis.com/jnl-up-j-tismir-files/journals/1/articles/43/submission/proof/43-1-1337-1-10-20200824.pdf). I'll make sure to at least add a comment that refers to this tomorrow.

Reply:
Done.